src/sleep/sleep.c

```diff
@@ -263,41 +263,83 @@ static int execute(
-        if (r < 0)
-                return log_error_errno(errno, "Error setting hibernate timer: %m");
+                last_capacity = r;
+                before_timestamp = now(CLOCK_BOOTTIME);
 
-        r = execute(sleep_config, SLEEP_SUSPEND, NULL);
-        if (r < 0)
-                return r;
 
-        r = fd_wait_for_event(tfd, POLLIN, 0);
-        if (r < 0)
-                return log_error_errno(r, "Error polling timerfd: %m");
-        if (!FLAGS_SET(r, POLLIN)) /* We woke up before the alarm time, we are done. */
-                return 0;
+                log_debug("Set timerfd wake alarm to estimate battery discharge rate for %s", FORMAT_TIMESPAN(suspend_interval, USEC_PER_SEC));
+
+                timespec_store(&ts.it_value, suspend_interval);
+
+                r = timerfd_settime(tfd, 0, &ts, NULL);
+                if (r < 0)
+                        return log_error_errno(errno, "Error setting battery estimate timer: %m");
+
+                r = execute(sleep_config, SLEEP_SUSPEND, NULL);
+                if (r < 0)
+                        return r;
+
+                r = fd_wait_for_event(tfd, POLLIN, 0);
+                if (r < 0)
+                        return log_error_errno(r, "Error polling timerfd: %m");
+
+                if (!FLAGS_SET(r, POLLIN))  { /* We woke up before the alarm time, calculate battery discharged and estimate rate. */
+                        after_timestamp = now(CLOCK_BOOTTIME);
+
+                        time_elapsed = (after_timestamp - before_timestamp) / USEC_PER_SEC;
+
+                        r = read_battery_capacity_percentage();
+                        if (r < 0)
+                                return log_error_errno(r, "Error fetching battery capacity percentage: %m");
+
+                        current_capacity = r;
+                        log_debug("%d%% battery discharged in %f seconds time",
+                                        last_capacity - current_capacity,
+                                        time_elapsed);
+
+                        estimated_discharge_rate = ((last_capacity - current_capacity) * 60 * 60) / time_elapsed;
+
+                        log_debug("Battery discharge rate is %d%% per hour", estimated_discharge_rate);
+
+                        return 0;
+                }
+
+                log_debug("Attempting to estimate battery after waking from %s hours sleep timer",
+                          FORMAT_TIMESPAN(suspend_interval, USEC_PER_HOUR));
+
+                /* If woken up after alarm time, calculate battery capacity and battery capacity_level and estimate discharge rate */
+                r = read_battery_capacity_percentage();
+                if (r < 0)
+                        return log_error_errno(r, "Error fetching battery capacity: %m");
+
+                current_capacity = r;
+                estimated_discharge_rate = (last_capacity - current_capacity) / (suspend_interval * USEC_PER_HOUR);
```

Comment:
this info should be used to adjust the next sleep time

Reply:
yes. it is used by storing in file and then reading it from there on next wakeup